core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java

```diff
@@ -460,7 +473,8 @@ protected R load(int n, Index editInPlace) {
      *      If non-null, update this data structure.
      *      Otherwise do a copy-on-write of {@link #index}
      */
-    protected synchronized R load(File dataDir, Index editInPlace) {
+    private R load(File dataDir, Index editInPlace) {
+        assert Thread.holdsLock(this);
```

Comment:
So IIUC this will throw `AssertionException` if any concurrent call is done to this method without holding the thread lock, right? If so, is this not going to - possibly - break current calls (not taking this into account)?